File: Problems/Algorithms/543. Diameter of Binary Tree/diameter_binary_tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        def height(node: Optional[TreeNode]) -> int:
            if node is None or (node.left is None and node.right is None):
                return 0

            left, right = height(node.left), height(node.right)
            return max(left, right) + 1

        if root is None:
            return 0
        if root.left is None:
            return max(height(root), self.diameterOfBinaryTree(root.right))
        if root.right is None:
            return max(height(root), self.diameterOfBinaryTree(root.left))
        return max(2 + height(root.left) + height(root.right), self.diameterOfBinaryTree(root.left), self.diameterOfBinaryTree(root.right))

File: Problems/Algorithms/543. Diameter of Binary Tree/diameter-binary-tree.cpp
#include <algorithm>

using namespace std;

// Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
private:
    int helper(TreeNode* node, int& diameter) {
        if (!node) {
            return 0;
        }

        int left = helper(node->left, diameter);
        int right = helper(node->right, diameter);

        diameter = max(diameter, left + right);
        return max(left, right) + 1;
    }

public:
    int diameterOfBinaryTree(TreeNode* root) {
        int ans = 0;
        helper(root, ans);
        return ans;
    }
};
